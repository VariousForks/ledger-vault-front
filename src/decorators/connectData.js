//@flow
import { connect } from "react-redux";
import { denormalize } from "normalizr";
import React, { Component } from "react";
import isEqual from "lodash/isEqual";
import { fetchData } from "../redux/modules/data";
import type { APISpec } from "../data/api-spec";

type FetchData = (api: APISpec, body?: Object) => Promise<*>;
// prettier-ignore
type PropsWithoutA<Props, A> = $Diff<Props, A & {
  fetchData?: FetchData,
  loading?: boolean
}>;
// prettier-ignore
type In<Props, S> = Class<React.Component<Props, S>>;
// prettier-ignore
type Out<Props, A> = Class<React.Component<PropsWithoutA<Props, A>>>;
type Opts<Props, A> = {
  api: A,
  propsToApiParams?: (props: *) => Object,
  RenderLoading?: (props: PropsWithoutA<Props, A>) => *,
  RenderError?: (props: PropsWithoutA<Props, A>, error: Error) => *
};

type State = {
  results: ?Object,
  error: ?Error,
  loading: boolean
};

const defaultOpts = {
  RenderLoading: () => null,
  RenderError: () => null,
  propsToApiParams: _ => null
};

const neverEnding: Promise<any> = new Promise(() => {});

export default <Props, A: { [_: string]: APISpec }, S>(
  Decorated: In<Props, S>,
  opts: Opts<Props, A>
): Out<Props, A> => {
  const { api, propsToApiParams, RenderLoading, RenderError } = {
    ...defaultOpts,
    ...opts
  };
  const apiKeys = Object.keys(api);
  const displayName = `connectData(${Decorated.displayName ||
    Decorated.name ||
    ""})`;

  class Clazz extends Component<*, *> {
    static displayName = displayName;

    apiParams: *;

    state: State = {
      // local state to keep the results of api queries (only keeping the minimal normalized version here. i.e. the ids)
      results: null,
      // when data gets potentially reloaded, this is a state to track that.
      // TODO: in the future we might move it to the store so if the data gets globally reloaded we can provide this too... it might be a bool per API
      loading: false,
      // if any of the api fails, we will have an error
      error: null
    };

    _unmounted = false;

    sync(apiParams: *) {
      const { dispatch } = this.props;
      this.apiParams = apiParams;
      this.setState({ error: null, loading: true });
      Promise.all(apiKeys.map(key => dispatch(fetchData(api[key], apiParams))))
        .then(all => {
          if (this._unmounted) return;
          const results = {};
          all.forEach((result, i) => {
            results[apiKeys[i]] = result;
          });
          this.setState({ results, loading: false });
        })
        .catch(error => {
          if (this._unmounted) return;
          this.setState({
            error,
            loading: false
          });
        });
    }

    /**
     * Perform an API call. body can be provided for POST apis
     * this can also be used to "refresh" data.
     */
    fetchData: FetchData = (api, body) =>
      this.props
        .dispatch(fetchData(api, this.apiParams, body))
        .then(
          result =>
            this._unmounted
              ? neverEnding
              : denormalize(
                  result,
                  api.responseSchema,
                  this.props.dataStore.entities
                )
        );

    componentWillMount() {
      this.sync(propsToApiParams(this.props));
    }

    componentWillUnmount() {
      this._unmounted = true;
    }

    componentWillReceiveProps(props: *) {
      const apiParams = propsToApiParams(props);
      if (!isEqual(apiParams, this.apiParams)) {
        this.sync(apiParams);
      }
    }

    render() {
      const { dataStore, ...props } = this.props;
      const { results, error, loading } = this.state;
      if (error) return <RenderError {...props} error={error} />;
      if (!results) return <RenderLoading {...props} />;
      const extraProps = {
        fetchData: this.fetchData,
        loading
      };
      apiKeys.forEach(key => {
        const data = denormalize(
          results[key],
          api[key].responseSchema,
          dataStore.entities
        );
        extraProps[key] = data;
      });
      return <Decorated {...props} {...extraProps} />;
    }
  }

  return connect(({ data }) => ({ dataStore: data }))(Clazz);
};
