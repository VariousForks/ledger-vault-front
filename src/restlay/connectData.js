//@flow
import { connect } from "react-redux";
import React, { Component } from "react";
import PropTypes from "prop-types";
import shallowEqual from "fbjs/lib/shallowEqual";
import isEqual from "lodash/isEqual";
import {
  executeQueryOrMutation,
  getQueryCacheResult,
  queryCacheIsFresh
} from "./dataStore";
import type { Store } from "./dataStore";
import Query from "./Query";
import Mutation from "./Mutation";
import type RestlayProvider from "./RestlayProvider";

export type RestlayEnvironment = {|
  commitMutation: <In, Res>(m: Mutation<In, Res>) => Promise<Res>,
  fetchQuery: <In, Res>(m: Query<In, Res>) => Promise<Res>,
  forceFetch: () => Promise<void>
  /* IDEA
  isReloadingData: (data: Object) => boolean,
  isOptimisticData: (data: Object) => boolean
  */
|};

type ConnectedProps = {|
  dataStore: Store,
  dispatch: Function
|};

type InjectedProps = {|
  restlay: RestlayEnvironment,
  reloading: boolean
|};

type ExtractQuery = <Q>(Class<Q>) => Q;
type ExtractQueryResult = <In, Out>(Class<Query<In, Out>>) => Out;
type ExtractQueryIn = <In, Out>(Class<Query<In, Out>>) => In;

// prettier-ignore
type InProps<Props, A> = $Supertype<InjectedProps & $ObjMap<A, ExtractQueryResult> & Props>;
type In<Props, A> =
  | Class<React$Component<InProps<Props, A>, any>>
  | ((props: InProps<Props, A>, ctx: any) => any);
// prettier-ignore
type Out<Props> = Class<React$Component<Props, any>>;

type ClazzProps<Props> = { ...ConnectedProps, ...Props };

export type ContextOverridableOpts<Props> = {
  // allow to implement the loading rendering. default is blank
  // prettier-ignore
  RenderLoading?: React$ComponentType<$Shape<{
    restlay: RestlayEnvironment
  } & Props>>,

  // allow to implement the error rendering. default is blank
  // prettier-ignore
  RenderError?: React$ComponentType<$Shape<{
    error: Error,
    restlay: RestlayEnvironment
  } & Props>>
};

type Opts<Props, A> = ContextOverridableOpts<Props> & {
  // an object of { [propName: string]: Class<Query> }
  queries?: A,

  // allow to pass parameters to the api uri function that will be used to generate api URL.
  propsToQueryParams?: (props: Props) => $Values<$ObjMap<A, ExtractQueryIn>>,

  // if a cached was defined on the Query, ignore it and make sure to reload the latest data
  forceFetch?: boolean,

  // allow the component to render if the api data was previously loaded
  optimisticRendering?: boolean
};

const defaultOpts = {
  queries: {},
  RenderLoading: () => null,
  RenderError: () => null,
  propsToQueryParams: _ => ({}),
  forceFetch: false,
  optimisticRendering: false
};

const extractInputProps = <Props>(props: ClazzProps<Props>): Props => {
  const { dataStore, dispatch, ...rest } = props; // eslint-disable-line no-unused-vars
  return rest;
};

const mapStateToProps = (state: Object): { dataStore: Store } => ({
  dataStore: state.data
});

export default function connectData<
  A: { [key: string]: Class<Query<any, any>> },
  Props: Object
>(Decorated: In<Props, A>, opts?: Opts<Props, A>): Out<Props> {
  type APIProps = $ObjMap<A, ExtractQueryResult>;

  const displayName = `connectData(${Decorated.displayName ||
    Decorated.name ||
    ""})`;

  // options that are not overridable by restlayProvider
  const { queries, propsToQueryParams, optimisticRendering, forceFetch } = {
    ...defaultOpts,
    ...opts
  };
  const queriesKeys: Array<$Keys<A>> = Object.keys(queries);

  type State = {
    catchedError: ?Error,
    apiError: ?Error,
    pending: boolean,
    data: ?APIProps
  };

  class Clazz extends Component<ClazzProps<Props>, State> {
    context: {
      restlayProvider: RestlayProvider
    };

    static displayName = displayName;
    static contextTypes = {
      restlayProvider: PropTypes.object.isRequired
    };

    state = {
      // this holds the componentDidCatch errors
      catchedError: null,
      apiError: null,
      // we need to wait a potential initial sync to not render initially with previous data if optimisticRendering is not asked
      pending: false,
      data: null
    };

    _options = {
      ...defaultOpts,
      ...this.context.restlayProvider.props.connectDataOptDefaults,
      ...opts
    };
    getOptions() {
      return this._options;
    }

    apiParams: ?Object = null;
    queriesInstances: ?$ObjMap<A, ExtractQuery>;
    _unmounted = false;

    updateQueryInstances(apiParams: Object) {
      const instances: $ObjMap<A, ExtractQuery> = {};
      queriesKeys.forEach(key => {
        const Q = queries[key];
        const query: Query<*, *> = new Q(apiParams);
        instances[key] = query;
      });
      this.queriesInstances = instances;
      this.apiParams = apiParams;
    }

    syncAPI(
      apiParams: Object,
      forceFetchMode: boolean = false
    ): Array<Promise<*>> {
      const { dataStore } = this.props;
      if (this.apiParams !== apiParams || !this.queriesInstances) {
        this.updateQueryInstances(apiParams);
      }
      const queriesInstances = this.queriesInstances;
      if (!queriesInstances) throw new Error("no queriesInstances in sync !?");
      // $FlowFixMe filter() don't seem to be flowtyped correctly
      const promises: Array<Promise<*>> = queriesKeys
        .map(key => {
          const query = queriesInstances[key];
          if (
            forceFetchMode ||
            forceFetch ||
            !queryCacheIsFresh(dataStore, query)
          ) {
            return this.execute(query);
          }
        })
        .filter(p => p);
      return promises;
    }

    syncAPI_id = 0;

    syncProps(
      props: ClazzProps<Props>,
      forceFetchMode: boolean = false
    ): Promise<*> {
      // FIXME can we simplify the code?

      let { pending, data } = this.state;
      const { queriesInstances } = this;
      const { dataStore } = props;
      let p: ?Promise<*>;

      const apiParams = propsToQueryParams(extractInputProps(props));
      if (
        !this.apiParams ||
        !isEqual(apiParams, this.apiParams) ||
        forceFetchMode
      ) {
        const syncId = ++this.syncAPI_id;
        const promises = this.syncAPI(apiParams, forceFetchMode);
        if (promises.length > 0) {
          pending = true;
          if (!this.state.pending) this.setState({ pending });
          p = Promise.all(promises).then(
            () => {
              if (this._unmounted || syncId !== this.syncAPI_id) return;
              this.setState({ pending: false }, () => {
                this.syncProps(this.props); // we need to sync again to make sure data is in sync
              });
            },
            apiError => {
              if (this._unmounted || syncId !== this.syncAPI_id) return;
              this.setState({ apiError, pending: false });
            }
          );
        }
      }

      const results = [];
      for (let key in queriesInstances) {
        const query = queriesInstances[key];
        const cache = getQueryCacheResult(dataStore, query);
        if (cache) {
          const { result } = cache;
          results.push({
            result,
            key,
            query
          });
        }
      }

      if (results.length === queriesKeys.length) {
        if (!pending || optimisticRendering) {
          const newData: APIProps = {};
          let nbOfChanges = 0;
          results.forEach(({ query, result, key }) => {
            let item = query.getResponse(result, dataStore);
            if (data && isEqual(item, data[key])) {
              // keep previous reference if deep equals
              item = data[key];
            } else {
              nbOfChanges++;
            }
            newData[key] = item;
          });
          if (nbOfChanges > 0) {
            // only set newData if it's actually new data^^
            this.setState({ data: newData });
          }
        }
      }

      return p || Promise.resolve();
    }

    execute<In, Out>(
      queryOrMutation: Query<In, Out> | Mutation<In, Out>
    ): Promise<Out> {
      return this.props.dispatch(
        executeQueryOrMutation(this.context.restlayProvider.props.network)(
          queryOrMutation
        )
      );
    }

    // This "environment" is a restlay prop that will be passed to the component.
    // the idea is to put everything in an object to not pollute props and introduce new things over time
    restlay: RestlayEnvironment = {
      commitMutation: <In, Res>(m: Mutation<In, Res>): Promise<Res> => {
        if (!(m instanceof Mutation)) {
          console.error(m);
          throw new Error(
            "invalid mutation provided in restlay.commitMutation"
          );
        }
        return this.execute(m);
      },
      fetchQuery: <In, Res>(query: Query<In, Res>): Promise<Res> => {
        if (!(query instanceof Query)) {
          console.error(query);
          throw new Error("invalid mutation provided in restlay.fetchQuery");
        }
        return this.execute(query);
      },
      forceFetch: () => this.syncProps(this.props, true).then(() => {})
    };

    componentWillMount() {
      this.syncProps(this.props);
    }

    componentWillUnmount() {
      this._unmounted = true;
    }

    componentWillReceiveProps(props: ClazzProps<Props>) {
      this.syncProps(props);
    }

    componentDidCatch(catchedError: Error) {
      this.setState({ catchedError });
    }

    shouldComponentUpdate(props: ClazzProps<Props>, state: State) {
      return (
        !shallowEqual(
          extractInputProps(this.props),
          extractInputProps(props)
        ) || !shallowEqual(this.state, state)
      );
    }

    render() {
      const { restlay } = this;
      const { data, apiError, catchedError, pending } = this.state;
      const props = extractInputProps(this.props);
      const error = catchedError || apiError;
      const { RenderError, RenderLoading } = this.getOptions();
      if (error) {
        return <RenderError {...props} error={error} restlay={restlay} />;
      }
      if (!data) {
        return <RenderLoading {...props} restlay={restlay} />;
      }
      return (
        <Decorated {...props} {...data} reloading={pending} restlay={restlay} />
      );
    }
  }

  return connect(mapStateToProps)(Clazz);
}
