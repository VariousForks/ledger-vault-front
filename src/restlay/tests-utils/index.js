//@flow
import React, { Component } from "react";
import { Provider } from "react-redux";
import { createStore, applyMiddleware, combineReducers } from "redux";
import thunk from "redux-thunk";
import RestlayProvider from "../RestlayProvider";
import { reducer } from "../dataStore";
import type { NetworkF } from "../dataStore";

export class NullComponent extends Component<{}> {
  render() {
    return null;
  }
}

export function flushPromises(): Promise<any> {
  return new Promise(resolve => setImmediate(resolve));
}

export const newStore = () =>
  createStore(
    combineReducers({
      data: reducer
    }),
    applyMiddleware(thunk)
  );

export const createRender = (
  network: NetworkF,
  connectDataOptDefaults?: Object,
  store?: * = newStore()
) => {
  const render = (children: React$Node) => (
    <Provider store={store}>
      <RestlayProvider
        network={network}
        connectDataOptDefaults={connectDataOptDefaults}
      >
        {children}
      </RestlayProvider>
    </Provider>
  );
  return render;
};

export const defer = () => {
  let resolve: () => void, reject: () => void;
  const promise: Promise<any> = new Promise((success, failure) => {
    resolve = success;
    reject = failure;
  });
  if (!resolve || !reject) throw new Error(); // never happen, to satisfy flow
  return { resolve, reject, promise };
};

// returns {network,tick}
// network function: can be passed to createRender
// tick function: needs to be called to redeem pending network calls. (also returns number of waiters)
// countTickWaiters function: gets the number of network call awaiting
export const networkFromMock = (
  syncMockFunction: Function
): {
  network: NetworkF,
  tick: () => number,
  countTickWaiters: () => number
} => {
  let tickD = defer();
  let tickWaiters = 0;
  const waitTick = () => {
    ++tickWaiters;
    return tickD.promise.then(() => {
      --tickWaiters;
    });
  };
  return {
    tick: () => {
      const nb = tickWaiters;
      tickD.resolve();
      tickD = defer();
      return nb;
    },
    countTickWaiters: () => tickWaiters,
    network: (uri, method, body) =>
      waitTick().then(() => syncMockFunction(uri, method, body))
  };
};

// Just a dumb test to ignore jest warning there is no tests
test("utils", () => {
  expect(createRender).toBeInstanceOf(Function);
});
