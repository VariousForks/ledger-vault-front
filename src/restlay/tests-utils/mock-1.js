//@flow
import type { NetworkF } from "../dataStore";
import Query from "../Query";
import Mutation from "../Mutation";
import { schema } from "normalizr";

type Animal = {
  id: string,
  name: string,
  age: number
};
const schemaAnimal = new schema.Entity("animals");

export class AnimalsQuery extends Query<void, Animal[]> {
  uri = "/animals";
  responseSchema = [schemaAnimal];
}

export class AddAnimalMutation extends Mutation<{ animal: Object }, Animal> {
  method = "POST";
  uri = "/animals";
  responseSchema = [schemaAnimal];
  getBody() {
    return this.props.animal;
  }
}

const genId = ((id: number) => () => "id_" + ++id)(0);

const animals: Animal[] = [
  {
    id: genId(),
    name: "max",
    age: 14
  },
  {
    id: genId(),
    name: "doge",
    age: 5
  }
];

export const syncNetwork = (uri: string, method: string, body: any): any => {
  if (method === "GET" && uri === "/animals") {
    return animals;
  }
  if (method === "POST" && uri === "/animals") {
    const animal = { id: genId(), ...body };
    animals.push(animal);
    return animal;
  }
  throw "notfound";
};
