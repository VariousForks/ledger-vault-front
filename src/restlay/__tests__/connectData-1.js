//@flow
import React, { Component } from "react";
import renderer from "react-test-renderer";
import connectData from "../connectData";
import { createRender, networkFromMock, flushPromises } from "../tests-utils";
import { syncNetwork, AnimalsQuery } from "../tests-utils/mock-1";

test("the mock-1 and utility functions works", () => {
  const net = networkFromMock(syncNetwork);
  const render = createRender(net.network);
  const inst = renderer.create(render(<div />));
  expect(inst.toJSON()).toMatchSnapshot();
  expect(net.countTickWaiters()).toBe(0);
  expect(net.tick()).toBe(0);
});

test("basic connectData component", async () => {
  const net = networkFromMock(syncNetwork);
  const render = createRender(net.network);
  class AnimalsR extends Component<*> {
    render() {
      const { animals } = this.props;
      return (
        <div>
          {animals.map(a => (
            <div key={a.id}>
              {a.name}: {a.age}
            </div>
          ))}
        </div>
      );
    }
  }
  const Animals = connectData(AnimalsR, {
    queries: {
      animals: AnimalsQuery
    }
  });
  const inst = renderer.create(render(<Animals />));
  expect(inst.toJSON()).toMatchSnapshot();
  expect(net.tick()).toBe(1);
  await flushPromises();
  expect(net.countTickWaiters()).toBe(0);
  expect(inst.toJSON()).toMatchSnapshot();
});
