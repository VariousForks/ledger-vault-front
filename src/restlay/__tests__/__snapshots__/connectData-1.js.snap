// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`basic connectData component 1`] = `null`;

exports[`basic connectData component 2`] = `
<div>
  <div>
    max
    : 
    14
  </div>
  <div>
    doge
    : 
    5
  </div>
</div>
`;

exports[`the mock-1 and utility functions works 1`] = `<div />`;
